Dockerfile

```diff
@@ -278,42 +278,75 @@ RUN echo "Pip version: ${PIP_VERSION}"
 
 RUN pip install --upgrade pip==${PIP_VERSION}
 
-# We are copying everything with airflow:airflow user:group even if we use root to run the scripts
+ARG AIRFLOW_REPO=apache/airflow
+ENV AIRFLOW_REPO=${AIRFLOW_REPO}
+
+ARG AIRFLOW_BRANCH=master
+ENV AIRFLOW_BRANCH=${AIRFLOW_BRANCH}
+
+ENV AIRFLOW_GITHUB_DOWNLOAD=https://raw.githubusercontent.com/${AIRFLOW_REPO}/${AIRFLOW_BRANCH}
+
+# We perform fresh dependency install at the beginning of each month from the scratch
+# This way every month we re-test if fresh installation from the scratch actually works
+# As opposed to incremental installations which does not upgrade already installed packages unless it
+# is required by setup.py constraints.
+ARG BUILD_MONTH
+
+# We get Airflow dependencies (no Airflow sources) from the master version of Airflow in order to avoid full
+# pip install layer cache invalidation when setup.py changes. This can be reinstalled from the
+# latest master by increasing PIP_DEPENDENCIES_EPOCH_NUMBER.
+RUN mkdir -pv ${AIRFLOW_SOURCES}/airflow/bin \
+ && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/setup.py >${AIRFLOW_SOURCES}/setup.py \
+ && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/setup.cfg >${AIRFLOW_SOURCES}/setup.cfg \
```

Comment:
You only need to build the docker image once and then this layer will be cached by docker. as long as BUILD_MONTH remains the same Docker will not invalidate this layer (It will not check if the master have changed or anything like that).
If you simply run 'docker build .' then BUILD_MONTH will be always empty and the build will remain in the cache and never try to download stuff from github.

If you use './hooks/build' , the BUILD_MONTH arg will change every month and every month it will re-download setup.py/setup/cfg etc. from master and reinstall all dependencies from the scratch. For that you will need to have access to internet (basically you need it for pip install anyway) - but it will only trigger when you actually run the build.


In most cases - if you make a change to setup.py you will have to run the build and then it will use cache until the COPY -- line below - where it will see that setup.py changed and it will run 'pip install' right after that. But then - you have to have access to internet to run pip install anyway. And again - it will only be needed if you run docker build which should not happen often (and you do not need it except if you rebase with other changes or change setup.py). Sources from your host will be mounted inside docker image so there is no need to rebuild it for local development at all.

However indeed decision on when to run docker build is not obvious and even checking if cache layer is invalidated can take some time so  I actually address it in the   #4932 which I will shortly open for proposal/voting. This is 'breeze' development environment based on this new image. There I implemented additional checking of md5sum on crucial files (like setup.py or Dockerfile) and when entering the environment with `breeze` command I will warn the user that the local environment needs updating (and the user will have option to rebuild or continue without it) - this way - once you run `breeze` at least once, you can continue using it locally without any internet access and you will get warned when you need to update.

Reply:
@dimberman - does this explanation help :)? Still requesting change on it ?

Reply:
Hey @dimberman :) ?

Reply:
Dan's on an extended vacation/travel right now (back soon-ish) so isn't around much.

I think I echo Dan's concerns here though -- this is an unusual pattern and means you can't build a new docker image if you are sans-connection or if Github is having problems.

What are we doing this rather than copying the files from the checkout?

Additionally even if we keep this pattern add `--fail` to the curl commands (As a rough rule I prefer long-form options in automated scripts.)

Reply:
It's really an optimisation only. If I copy the setup.py from local context, it will invalidate all the subsequent layers when the setup.py changes - which means that on setup.py change we always reinstall everything from scratch. 

This is both good and bad. Good that pip install is from the scratch. Bad - that it takes a lot of time every time you add/modify even a single dependency. 

For sure when we make a production image we should skip that step and always reinstall from the scratch. But for CI images and also following steps like breeze or pre-commit hooks, I think it makes perfect sense to optimise for speed of building. 

In the pattern I propose - we already have all "fresh" dependencies installed from master - and only as next step we upgrade/downgrade as needed according to the actual "setup.py" we have in sources. This is so much faster than installing from the scratch - build time in this case goes down from like 4 minutes to less than 1. This adds up in CI builds. If you have setup.py change - until the change is merged into master (and DockerHub build completes), it will re-run the pip install with every PR rebase.

Also the once-a-month full reinstall (which is now automated) is also good - it will keep us from accumulating incremental changes.

I think this optimisation is a bit novel/unusual indeed but so far I found no problems with that - you anyhow need access to PIP to install it.

What I can do as well, Is I can make failures in this step (and the following pip-install) non-terminal. This way even if - for example - github is not accessible, it will simply silently fail. This is really an idempotent step - so the next step's 'pip install' will work fine regardless if this "optimisation" one succeeds.

Reply:
This feels like another image layer - rather than being in here. Would that complicate things needlessly?

Is the goal here to speed up local dev, or does this help on PRs too?  (If the goal is to help local dev then setting a pip cache dir to a volume mount to cache wheels/downloads go a long way towards speeding it up, without this odd pattern. I'm not sure what I don't like about it.

I also wonder if we could do this differently:

```
RUN pip install --no-use-pep517 "pip install 'https://github.com/apache/airflow/archive/master.tar.gz#egg=apache-airflow[aws]'[${AIRFLOW_EXTRAS}]" \
  && pip uninstall apache-airflow
```

Being able to do it in a single command/layer makes me happier.

Reply:
It's not only for local development. It's also for CI builds. Just to explain what's going on here:

The only way to move setup.py to docker context locally is to run COPY or ADD operation. Whenever setup.py changes and is part of the COPY/ADD then it means that Dockerfile gets invalidated from that point on.

The case for CI this pattern solves is when your PR adds a new dependency.

Without it - when you make A PR with dependency added in setup.py - the COPY/ADD causes that whole 'pip install' layer gets invalidated and re-installed from the scratch. This means that every job (pylint/mypy/tests etc.) will have to first perform the whole `pip install` from the scratch adding 4-5 minutes to the build time. 

It is a bit strange pattern indeed. Simply because Docker build does not support this pattern yet :(. Howewer this change is not for ever and Docker/moby guys realise that. There are layer types (including cache specifically foreseen for that kind of behaviour) in the upcoming buildkit: https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/experimental.md#run---mounttypecache  - but it will likely take months to release stable version with buildkit that we can use (but when they will, I am super happy to update the docker image to follow it).

Also we cannot do the single line you mentioned. In order to do that we would have to add `COPY .` (which we do eventually few lines below). Doing it will make it even worse - every single change to airflow sources will invalidate everything and will force 'pip install' to run from the scratch basically with every commit. That's why I separately COPY setup.py, setup.cfg bin/airflow first - because they do not change that often and they are needed to run the first pip install.

I see your concern. While I understand it's not perfect solution, I think it is really good for CI optimisations until we have buildkit. What I can do instead - I can make those steps already controlled with "OPTIMISE_FOR_CI_BUILD" argument and skip those steps entirely when it is not set to "true" (which  I will do in the CI/breeze/pylint/mypy cases). Then the "non-CI-build-optimised" images will run just single "pip install" - the last one in the Dockerfile :). Also that will make it obvious that this is an optimisation rather than integral part of the "basic" Dockerfile.

Would that be a good solution ?

Reply:
I don't see the caching difference between these two.

```Dockerfile
ARG BUILD_MONTH
RUN pip install --no-use-pep517 "pip install "https://github.com/apache/airflow/archive/master.tar.gz#egg=apache-airflow[${AIRFLOW_EXTRAS}]" \
  && pip uninstall --yes apache-airflow
```

and 

```Dockerfile
ARG BUILD_MONTH
RUN mkdir -pv ${AIRFLOW_SOURCES}/airflow/bin \
 && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/setup.py >${AIRFLOW_SOURCES}/setup.py \
 && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/setup.cfg >${AIRFLOW_SOURCES}/setup.cfg \
 && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/airflow/version.py >${AIRFLOW_SOURCES}/airflow/version.py \
 && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/airflow/__init__.py >${AIRFLOW_SOURCES}/airflow/__init__.py \
 && curl -L ${AIRFLOW_GITHUB_DOWNLOAD}/airflow/bin/airflow >${AIRFLOW_SOURCES}/airflow/bin/airflow

# Airflow Extras installed
ARG AIRFLOW_EXTRAS="all"
ENV AIRFLOW_EXTRAS=${AIRFLOW_EXTRAS}
RUN echo "Installing with extras: ${AIRFLOW_EXTRAS}."

RUN pip install --no-use-pep517 -e ".[${AIRFLOW_EXTRAS}]"
```

In either case the end result of these instructions is airflow deps installed, based on master deps once per month. Docker doesn't "know" about the content of the files when they are added by a `RUN` command - so changing the content of setup.py (etc.) in master will not invalidate the cache in either mechanism.

Reply:
Ah. OK. I see. I misread it. It is indeed much simpler. I did not even know you can do it this way.  It's cool :).

Reply:
It's now don this simple way and it works perfectly well. Good one @ashb